### PR TITLE
Updated squeeze and squeeze.dim batching rules to new style [#27]

### PR DESCRIPTION
**Pull request by laurencer**
Addresses #27 by updating `squeeze` and `squeeze.dim` to the new batching rules style while also adding special casing for scalar arrays to match underlying PyTorch behavior.

### CODE REVIEW

**Comment by laurencer** [comment]
@zou3519 - let me know what you think. I might try to tidy up the tests a bit so there's less copied code but would be good to get your initial review/thoughts on the approach.

File: functorch/csrc/BatchingRegistrations.cpp [hunk 1]
```diff
@@ -1372,8 +1415,8 @@ TORCH_LIBRARY_IMPL(aten, FT_BATCHED_KEY, m) {
   m.impl("slice.Tensor", slice_batching_rule);
   m.impl("split.Tensor", split_batching_rule);
   m.impl("split_with_sizes", split_with_sizes_batching_rule);
-  m.impl("squeeze", squeeze_batching_rule);
-  m.impl("squeeze.dim", squeeze_dim_batching_rule);
+  VMAP_SUPPORT("squeeze", squeeze_batching_rule);
+  VMAP_SUPPORT("squeeze.dim", squeeze_dim_batching_rule);
```
Comment (zou3519) [inline functorch/csrc/BatchingRegistrations.cpp:1419]: Let's move these functions to `BatchRulesView.cpp` -- squeeze is a view function and BatchingRegistrations.cpp is meant for the old-style batching rules (it's not documented as so though)

File: test/test_vmap.py [hunk 2]
```diff
@@ -2106,16 +2106,41 @@ def test_slice(self):
...
+            test(op, (torch.rand(B0, 0, 1, 5, 1),))
+            test(op, (torch.rand(B0, 1, 1, 1, 1),))
+            test(vmap(op), (torch.rand(B0, B1, 1, 3, 4),))
+            test(vmap(op), (torch.rand(B1, 1, B0, 4, 5),), in_dims=2)
+
+        verify_behavior(torch.squeeze)
+        verify_behavior(lambda x: torch.squeeze(x, dim=1), min_size=2)
+        verify_behavior(lambda x: torch.squeeze(x, dim=-1), min_size=2)
+        verify_behavior(lambda x: torch.squeeze(x, dim=-2), min_size=3)
+
```
Comment (zou3519) [inline test/test_vmap.py:2130]: I think we can uncomment out https://github.com/facebookresearch/functorch/blob/b73213bad5c5c6aa5eeaa96a048bf1163f4cdb40/test/test_vmap.py#L2908. PyTorch has an operator database and we have some autogenerated tests for each operator in that database. Our autogenerated tests for vmap aren't as comprehensive as manually writing tests yet (so the tests you wrote are still helpful!)

File: test/test_vmap.py [hunk 3]
```diff
@@ -2106,16 +2106,41 @@ def test_slice(self):
...
-        B0, B1 = 1, 11
-        test(op, (torch.rand(B0),))
-        test(op, (torch.rand(B0, 3, 5),))
-        test(op, (torch.rand(1, B0, 5),), in_dims=1)
-        test(op, (torch.rand(B0, 0, 1, 5, 1),))
-        test(op, (torch.rand(B0, 1, 1, 1, 1),))
-        test(vmap(op), (torch.rand(B0, B1, 1),))
-        test(vmap(op), (torch.rand(B1, 1, B0),), in_dims=2)
+        assert vmap(op)(x).equal(torch.stack([op(xi) for xi in x.unbind(0)]))
+        test(op, (x,))
```
Comment (zou3519) [inline test/test_vmap.py:2138]: `_vmap_view_test` should already check `assert vmap(op)(x).equal(torch.stack([op(xi) for xi in x.unbind(0)]))`

File: test/test_vmap.py [hunk 4]
```diff
@@ -2106,16 +2106,41 @@ def test_slice(self):
              (torch.rand(3, 5, B0, B1, B2),), in_dims=2)
 
     def test_squeeze(self):
+        def verify_behavior(op, min_size=1):
```
Comment (zou3519) [inline test/test_vmap.py:2109]: By size do you mean "number of dimensions"? If so we should name this `min_rank` or `min_ndim`.

File: functorch/csrc/BatchingRegistrations.cpp [hunk 5]
```diff
@@ -448,40 +448,83 @@ Tensor& zero_inplace_batching_rule(Tensor &self) {
...
-  int64_t num_batch_dims = self_physical.numBatchDims();
-  squeezed_sizes.insert(
-      squeezed_sizes.end(),
-      physical_sizes.begin(),
-      physical_sizes.begin() + num_batch_dims);
-  for (auto it = physical_sizes.begin() + num_batch_dims; it != physical_sizes.end(); ++it) {
-    if (*it != 1) {
-      squeezed_sizes.push_back(*it);
+std::tuple<Tensor, optional<int64_t>> squeeze_batching_rule(const Tensor& self, optional<int64_t> bdim) {
+  if (bdim) {
```
Comment (zou3519) [inline functorch/csrc/BatchingRegistrations.cpp:452]: I forgot to mention this, but `bdim` is never nullopt in an operator that accepts a single Tensor. If bdim is nullopt, then the batching rule doesn't actually get invoked; PyTorch dispatcher ends up not dispatching to the Batched implementation of things. Because of this we can add `TORCH_INTERNAL_ASSERT(bdim.has_value())` and delete the branch where bdim is nullopt.

File: functorch/csrc/BatchingRegistrations.cpp [hunk 6]
```diff
@@ -448,40 +448,83 @@ Tensor& zero_inplace_batching_rule(Tensor &self) {
...
+    if (self.dim() == 1) {
+      // Special case for scalar arrays to replicate PyTorch behavior.
+      return std::make_tuple<Tensor, optional<int64_t>>(
```
Comment (zou3519) [inline functorch/csrc/BatchingRegistrations.cpp:455]: nit: I think `make_tuple` can deduce the type of its template so no need to write it out: `std::make_tuple(...)`

File: functorch/csrc/BatchingRegistrations.cpp [hunk 7]
```diff
@@ -448,40 +448,83 @@ Tensor& zero_inplace_batching_rule(Tensor &self) {
...
+          // Only increment for the dimensions that will be kept in the output.
+          if (before_batch_idx) {
+            ++new_batch_idx;
+          }
+        }
+        ++original_idx;
+      }
+
+      auto result = self.view(squeezed_sizes);
+      return std::make_tuple<Tensor, optional<int64_t>>(std::move(result), c10::optional<int64_t>(new_batch_idx));
```
Comment (zou3519) [inline functorch/csrc/BatchingRegistrations.cpp:486]: nit: I think `make_tuple` can deduce the type of its template so no need to write it out: `std::make_tuple(...)`

File: functorch/csrc/BatchingRegistrations.cpp [hunk 8]
```diff
@@ -448,40 +448,83 @@ Tensor& zero_inplace_batching_rule(Tensor &self) {
...
+  } else {
+    // The array does not have a batch dimension so we can return using the underlying squeeze function.
+    return std::make_tuple<Tensor, optional<int64_t>>(self.squeeze(), c10::optional<int64_t>());
+  }
+}
+
+std::tuple<Tensor, optional<int64_t>> squeeze_dim_batching_rule(const Tensor& self, optional<int64_t> bdim, int64_t dim) {
+  if (bdim) {
+    if (self.dim() == 1) {
+      // Special case for scalar arrays to replicate PyTorch behavior.
```
Comment (zou3519) [inline functorch/csrc/BatchingRegistrations.cpp:497]: We should TORCH_CHECK that `dim == 0` here. PyTorch throws an error when that is not the case while squeezing a scalar:
```
x = torch.tensor(1.)
x.squeeze(1)
```

File: functorch/csrc/BatchingRegistrations.cpp [hunk 9]
```diff
@@ -448,40 +448,83 @@ Tensor& zero_inplace_batching_rule(Tensor &self) {
...
+      return std::make_tuple<Tensor, optional<int64_t>>(
+        std::move(self.alias()),
+        std::move(bdim)
+      );
+    } else {
+      // Calculate the proper offset if dim is negative.
+      auto actual_dim = dim;
+      if (dim < 0) {
+        actual_dim = self.dim() + dim - 1;
+      }
```
Comment (zou3519) [inline functorch/csrc/BatchingRegistrations.cpp:507]: We can use maybe_wrap_dim here: https://github.com/pytorch/pytorch/blob/7c588d5d00177230ac98bd92510e39bd0ce6f2da/aten/src/ATen/WrapDimUtils.h#L9-L11

I think the usage is `actual_dim = maybe_wrap_dim(dim, self.dim())` but there might be a off-by-one error in there

File: functorch/csrc/BatchingRegistrations.cpp [hunk 10]
```diff
@@ -448,40 +448,83 @@ Tensor& zero_inplace_batching_rule(Tensor &self) {
...
-      squeezed_sizes.push_back(*it);
+std::tuple<Tensor, optional<int64_t>> squeeze_batching_rule(const Tensor& self, optional<int64_t> bdim) {
+  if (bdim) {
+    if (self.dim() == 1) {
+      // Special case for scalar arrays to replicate PyTorch behavior.
+      return std::make_tuple<Tensor, optional<int64_t>>(
+        std::move(self.alias()),
+        std::move(bdim)
+      );
+    } else {
```
Comment (zou3519) [inline functorch/csrc/BatchingRegistrations.cpp:459]: nit: this is a style thing, but we can reduce the levels of indentation:
```
if (self.dim() == 1) {
   return ...;
}
// handle the other case here
```

File: functorch/csrc/BatchingRegistrations.cpp [hunk 11]
```diff
@@ -448,40 +448,83 @@ Tensor& zero_inplace_batching_rule(Tensor &self) {
...
-  for (auto it = physical_sizes.begin() + num_batch_dims; it != physical_sizes.end(); ++it) {
-    if (*it != 1) {
-      squeezed_sizes.push_back(*it);
+std::tuple<Tensor, optional<int64_t>> squeeze_batching_rule(const Tensor& self, optional<int64_t> bdim) {
+  if (bdim) {
+    if (self.dim() == 1) {
+      // Special case for scalar arrays to replicate PyTorch behavior.
+      return std::make_tuple<Tensor, optional<int64_t>>(
+        std::move(self.alias()),
+        std::move(bdim)
```
Comment (zou3519) [inline functorch/csrc/BatchingRegistrations.cpp:457]: Do we need std::move for these? (I'm not very good with C++ so I don't know and the rest of this repo has not been using std::move this way)

**Review by zou3519 — COMMENTED** [review]
This looks pretty good! I had some minor comments

**Comment by laurencer** [comment]
@zou3519 - thanks for the review. I pushed the changes that address your feedback!

**Review by zou3519 — APPROVED** [review]
LGTM, thank you @laurencer!
